File: Daily_Challenge/20221128_LC_380.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.28

class RandomizedSet:

    def __init__(self):
        self.nums_map = {}
        self.nums_list = []

    def insert(self, val: int) -> bool:
        result = False
        if val not in self.nums_map:
            result = True
        if result:
            self.nums_map[val] = len(self.nums_list)
            self.nums_list.append(val)
        return result

    def remove(self, val: int) -> bool:
        result = True
        if val not in self.nums_map:
            result = False
        if result:
            idx = self.nums_map[val]
            last_val = self.nums_list[-1]
            self.nums_list[idx] = last_val
            self.nums_list.pop()
            self.nums_map[last_val] = idx
            del self.nums_map[val]
        return result
    def getRandom(self) -> int:
        idx = random.randint(0, len(self.nums_list) - 1)
        return self.nums_list[idx]

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: Daily_Challenge/20221128_LC_380.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.28


class RandomizedSet {
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (nums_map.count(val)){
            return false;
        }
        nums_map[val] = nums_list.size();
        nums_list.push_back(val);
        return true;
    }
    
    bool remove(int val) {
        if (!nums_map.count(val)){
            return false;
        }
        // swap the to be removed element with the last element, then pop the last element
        int idx = nums_map[val];
        nums_map[nums_list.back()] = idx;
        nums_map.erase(val);
        swap(nums_list[idx], nums_list.back());
        nums_list.pop_back();
        return true;
    }
    
    int getRandom() {
        int idx = rand() % nums_list.size();
        return nums_list[idx];
    }
private:
    unordered_map<int, int> nums_map;
    vector<int> nums_list;
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */